libp2p/transports/tcptransport.nim

```diff
@@ -12,262 +12,325 @@
-    raise exc
-  except CatchableError as exc:
-    info "Unexpected error accepting connection", exc = exc.msg
-    raise exc
+    let remote =
+      try:
+        transp.remoteAddress
+      except TransportOsError as exc:
+        # The connection had errors / was closed before `await` returned control
+        await transp.closeWait()
+        debug "Cannot read remote address", exc = exc.msg
+        return nil
 
-method dial*(
-  self: TcpTransport,
-  hostname: string,
-  address: MultiAddress,
-  peerId: Opt[PeerId] = Opt.none(PeerId)): Future[Connection] {.async.} =
-  ## dial a peer
-  ##
+    let observedAddr =
+      MultiAddress.init(remote).expect("Can initialize from remote address")
+    self.connHandler(transp, Opt.some(observedAddr), Direction.In)
 
-  trace "Dialing remote peer", address = $address
-  let transp =
-    if self.networkReachability == NetworkReachability.NotReachable and self.addrs.len > 0:
-      self.clientFlags.incl(SocketFlags.ReusePort)
-      await connect(address, flags = self.clientFlags, localAddress = Opt.some(self.addrs[0]))
-    else:
-      await connect(address, flags = self.clientFlags)
-
-  try:
-    let observedAddr = MultiAddress.init(transp.remoteAddress).tryGet()
-    return await self.connHandler(transp, Opt.some(observedAddr), Direction.Out)
-  except CatchableError as err:
-    await transp.closeWait()
-    raise err
+  impl(self)
 
-method handles*(t: TcpTransport, address: MultiAddress): bool {.gcsafe.} =
+method dial*(
+    self: TcpTransport,
+    hostname: string,
+    address: MultiAddress,
+    peerId: Opt[PeerId] = Opt.none(PeerId),
+): Future[Connection] =
+  ## dial a peer
+  proc impl(
+      self: TcpTransport, hostname: string, address: MultiAddress, peerId: Opt[PeerId]
+  ): Future[Connection] {.async: (raises: [transport.TransportError, CancelledError]).} =
+    if self.stopping:
+      raise newTransportClosedError()
+
+    let ta = initTAddress(address).valueOr:
+      raise (ref TcpTransportError)(msg: "Unsupported address: " & $address)
+
+    trace "Dialing remote peer", address = $address
+    let transp =
+      try:
+        await(
+          if self.networkReachability == NetworkReachability.NotReachable and
+              self.addrs.len > 0:
+            let local = initTAddress(self.addrs[0]).expect("self address is valid")
+            self.clientFlags.incl(SocketFlags.ReusePort)
+            connect(ta, flags = self.clientFlags, localAddress = local)
+          else:
+            connect(ta, flags = self.clientFlags)
+        )
+      except CancelledError as exc:
+        raise exc
+      except CatchableError as exc:
+        raise (ref TcpTransportError)(msg: exc.msg, parent: exc)
+
+    # If `stop` is called after `connect` but before `await` returns, we might
+    # end up with a race condition where `stop` returns but not all connections
+    # have been closed - we drop connections in this case in order not to leak
+    # them
+    if self.stopping:
```

Comment:
This pattern is quite tricky and it's potentially unreliable to delegate to devs the responsibility to remember about those cases. Is there a way to automate it? Something happens when `stop` is called that cancels all pending awaited futures and the necessary cleanups happen in an `except` block.

Reply:
We could add those futures in a seq in the transport when they are created, and make them remove themselves when they complete. When we stop the transport we cancel all the futures in the seq.

Reply:
broadly, when we transition to 2.0 and the new gc, it will be possible to make this more ergonomic - until then, it's simply a pattern to learn and repeat